libtransact/src/contract/context/key_value.rs

```diff
@@ -0,0 +1,1310 @@
+        let mut new_entry_lists = Vec::new();
+        let mut delete_lists = Vec::new();
+        key_map.iter().for_each(|(nkey, addr)| {
+            // Fetching the StateEntryList at the corresponding address
+            if let Some(list) = state_entry_lists.get(addr) {
+                // The StateEntry objects will be filtered out of the StateEntryList if it has the
+                // normalized key. This normalized key is added to a list of successfully filtered
+                // entries to be returned.
+                if list.contains(nkey.to_string()) {
+                    let filtered = list
+                        .entries()
+                        .to_vec()
+                        .into_iter()
+                        .filter(|e| e.normalized_key() != nkey)
+                        .collect::<Vec<StateEntry>>();
+                    if filtered.is_empty() {
+                        delete_lists.push(addr.to_string());
+                    } else {
+                        new_entry_lists.push((addr.to_string(), filtered));
+                    }
+                    deleted_keys.push(nkey.to_string());
+                }
+            }
+        });
+        // Delete any StateEntryLists that have an empty list of entries
+        self.context.delete_state_entries(delete_lists.as_slice())?;
+        // Setting the newly filtered StateEntryLists into state using the internal context
+        self.context.set_state_entries(
+            new_entry_lists
+                .iter()
+                .map(|(addr, filtered_list)| {
+                    let new_entry_list = StateEntryListBuilder::new()
+                        .with_state_entries(filtered_list.to_vec())
+                        .build()
+                        .map_err(|err| KeyValueContextError::ProtocolBuildError(Box::new(err)))?;
+                    Ok((addr.to_string(), new_entry_list.into_bytes()?))
+                })
+                .collect::<Result<Vec<(String, Vec<u8>)>, KeyValueContextError>>()?,
+        )?;
+
+        Ok(deleted_keys)
+    }
+
+    /// add_receipt_data adds a blob to the execution result for this transaction.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - the data to add
+    pub fn add_receipt_data(&self, data: Vec<u8>) -> Result<(), KeyValueContextError> {
+        Ok(self.context.add_receipt_data(data)?)
+    }
+
+    /// add_event adds a new event to the execution result for this transaction.
+    ///
+    /// # Arguments
+    ///
+    /// * `event_type` -  A globally unique event identifier that is used to subscribe to events.
+    ///         It describes what, in general, has occured.
+    /// * `attributes` - Additional information about the event that is transparent to the
+    ///          validator. Attributes can be used by subscribers to filter the type of events
+    ///          they receive.
+    /// * `data` - Additional information about the event that is opaque to the validator.
```

Comment:
Mentions of the "validator" are awkward in Transact, which doesn't take a position on how the library is used (Sawtooth validator would be just one case).

Reply:
Reworded with same suggestion above used for the `add_event` method

Reply:
Reworded with same suggestion above used for the `add_event` method